File: binding/include/libqstandardpaths.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qstandardpaths.html
pub const qstandardpaths = struct {

    /// New constructs a new QStandardPaths object.
    pub fn New(other: ?*anyopaque) QtC.QStandardPaths {
        return qtc.QStandardPaths_new(@ptrCast(other));
    }


    /// New2 constructs a new QStandardPaths object.
    pub fn New2(param1: ?*anyopaque) QtC.QStandardPaths {
        return qtc.QStandardPaths_new2(@ptrCast(param1));
    }

    /// CopyAssign shallow copies `other` into `self`.
    pub fn CopyAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QStandardPaths_CopyAssign(@ptrCast(self), @ptrCast(other));
    }


    pub fn WritableLocation(typeVal: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QStandardPaths_WritableLocation(@intCast(typeVal));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qstandardpaths.WritableLocation: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn StandardLocations(typeVal: i32) []const u8 {
        return qtc.QStandardPaths_StandardLocations(@intCast(typeVal));
    }

    pub fn Locate(typeVal: i32, fileName: []const u8, allocator: std.mem.Allocator) []const u8 {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
const _str = qtc.QStandardPaths_Locate(@intCast(typeVal), fileName_str);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qstandardpaths.Locate: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn LocateAll(typeVal: i32, fileName: []const u8) []const u8 {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
return qtc.QStandardPaths_LocateAll(@intCast(typeVal), fileName_str);
    }

    pub fn DisplayName(typeVal: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QStandardPaths_DisplayName(@intCast(typeVal));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qstandardpaths.DisplayName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn FindExecutable(executableName: []const u8, allocator: std.mem.Allocator) []const u8 {
        const executableName_str = qtc.libqt_string{
    .len = executableName.len,
    .data = executableName.ptr,
};
const _str = qtc.QStandardPaths_FindExecutable(executableName_str);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qstandardpaths.FindExecutable: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetTestModeEnabled(testMode: bool) void {
        qtc.QStandardPaths_SetTestModeEnabled(testMode);
    }

    pub fn IsTestModeEnabled() bool {
        return qtc.QStandardPaths_IsTestModeEnabled();
    }

    pub fn Locate3(typeVal: i32, fileName: []const u8, options: i32, allocator: std.mem.Allocator) []const u8 {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
const _str = qtc.QStandardPaths_Locate3(@intCast(typeVal), fileName_str, @intCast(options));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qstandardpaths.Locate3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn LocateAll3(typeVal: i32, fileName: []const u8, options: i32) []const u8 {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
return qtc.QStandardPaths_LocateAll3(@intCast(typeVal), fileName_str, @intCast(options));
    }

    pub fn FindExecutable2(executableName: []const u8, paths: []const u8, allocator: std.mem.Allocator) []const u8 {
        const executableName_str = qtc.libqt_string{
    .len = executableName.len,
    .data = executableName.ptr,
};
const _str = qtc.QStandardPaths_FindExecutable2(executableName_str, @ptrCast(paths));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qstandardpaths.FindExecutable2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }
};

/// https://doc.qt.io/qt-6/standardpaths.html#types
pub const enums = struct {
    pub const StandardLocation = enum {
        pub const DesktopLocation: i32 = 0;
        pub const DocumentsLocation: i32 = 1;
        pub const FontsLocation: i32 = 2;
        pub const ApplicationsLocation: i32 = 3;
        pub const MusicLocation: i32 = 4;
        pub const MoviesLocation: i32 = 5;
        pub const PicturesLocation: i32 = 6;
        pub const TempLocation: i32 = 7;
        pub const HomeLocation: i32 = 8;
        pub const AppLocalDataLocation: i32 = 9;
        pub const CacheLocation: i32 = 10;
        pub const GenericDataLocation: i32 = 11;
        pub const RuntimeLocation: i32 = 12;
        pub const ConfigLocation: i32 = 13;
        pub const DownloadLocation: i32 = 14;
        pub const GenericCacheLocation: i32 = 15;
        pub const GenericConfigLocation: i32 = 16;
        pub const AppDataLocation: i32 = 17;
        pub const AppConfigLocation: i32 = 18;
        pub const PublicShareLocation: i32 = 19;
        pub const TemplatesLocation: i32 = 20;
        pub const StateLocation: i32 = 21;
        pub const GenericStateLocation: i32 = 22;
    };

    pub const LocateOption = enum {
        pub const LocateFile: i32 = 0;
        pub const LocateDirectory: i32 = 1;
    };

};

File: binding/src/libqstandardpaths.cpp
#include <QStandardPaths>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qstandardpaths.h>
#include "libqstandardpaths.h"
#include "libqstandardpaths.hxx"

QStandardPaths* QStandardPaths_new(const QStandardPaths* other) {
	 return new QStandardPaths(*other);
}

QStandardPaths* QStandardPaths_new2(const QStandardPaths* param1) {
	 return new QStandardPaths(*param1);
}

void QStandardPaths_CopyAssign(QStandardPaths* self, QStandardPaths* other) {
    *self = *other;
}

libqt_string QStandardPaths_WritableLocation(int typeVal) {
	QString _ret = QStandardPaths::writableLocation(static_cast<QStandardPaths::StandardLocation>(typeVal));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_list QStandardPaths_StandardLocations(int typeVal) {
	auto _ret = QStandardPaths::standardLocations(static_cast<QStandardPaths::StandardLocation>(typeVal));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

libqt_string QStandardPaths_Locate(int typeVal, const libqt_string fileName) {
	QString _ret = QStandardPaths::locate(static_cast<QStandardPaths::StandardLocation>(typeVal), QString::fromUtf8(fileName.data, fileName.len));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_list QStandardPaths_LocateAll(int typeVal, const libqt_string fileName) {
	auto _ret = QStandardPaths::locateAll(static_cast<QStandardPaths::StandardLocation>(typeVal), QString::fromUtf8(fileName.data, fileName.len));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

libqt_string QStandardPaths_DisplayName(int typeVal) {
	QString _ret = QStandardPaths::displayName(static_cast<QStandardPaths::StandardLocation>(typeVal));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QStandardPaths_FindExecutable(const libqt_string executableName) {
	QString _ret = QStandardPaths::findExecutable(QString::fromUtf8(executableName.data, executableName.len));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QStandardPaths_SetTestModeEnabled(bool testMode) {
	QStandardPaths::setTestModeEnabled(testMode);
}

bool QStandardPaths_IsTestModeEnabled() {
	return QStandardPaths::isTestModeEnabled();
}

libqt_string QStandardPaths_Locate3(int typeVal, const libqt_string fileName, int options) {
	QString _ret = QStandardPaths::locate(static_cast<QStandardPaths::StandardLocation>(typeVal), QString::fromUtf8(fileName.data, fileName.len), static_cast<QFlags<QStandardPaths::LocateOption>>(options));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_list QStandardPaths_LocateAll3(int typeVal, const libqt_string fileName, int options) {
	auto _ret = QStandardPaths::locateAll(static_cast<QStandardPaths::StandardLocation>(typeVal), QString::fromUtf8(fileName.data, fileName.len), static_cast<QFlags<QStandardPaths::LocateOption>>(options));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

libqt_string QStandardPaths_FindExecutable2(const libqt_string executableName, const libqt_list paths) {
	QString _ret = QStandardPaths::findExecutable(QString::fromUtf8(executableName.data, executableName.len), QList<QString>());
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

